deepchecks/nlp/checks/train_test_validation/property_drift.py

```diff
@@ -0,0 +1,270 @@
+        "EMD" for Earth Mover's Distance (EMD), "KS" for Kolmogorov-Smirnov (KS).
+    categorical_drift_method: str, default: "cramers_v"
+        decides which method to use on categorical variables. Possible values are:
+        "cramers_v" for Cramer's V, "PSI" for Population Stability Index (PSI).
+    ignore_na: bool, default True
+        For categorical columns only. If True, ignores nones for categorical drift. If False, considers none as a
+        separate category. For numerical columns we always ignore nones.
+    aggregation_method: Optional[str], default: 'l3_weighted'
+        {feature_aggregation_method_argument:2*indent}
+    min_samples : int , default: 10
+        Minimum number of samples required to calculate the drift score. If there are not enough samples for either
+        train or test, the check will return None for that feature. If there are not enough samples for all properties,
+        the check will raise a ``NotEnoughSamplesError`` exception.
+    n_samples : int , default: 100_000
+        Number of samples to use for drift computation and plot.
+    random_state : int , default: 42
+        Random seed for sampling.
+    """
+
+    def __init__(
+        self,
+        properties: t.Union[t.Hashable, t.List[Hashable], None] = None,
+        ignore_properties: t.Union[Hashable, t.List[Hashable], None] = None,
+        n_top_properties: int = 5,
+        margin_quantile_filter: float = 0.025,
+        max_num_categories_for_drift: t.Optional[int] = None,
+        min_category_size_ratio: float = 0.01,
+        max_num_categories_for_display: int = 10,
+        show_categories_by: str = 'largest_difference',
+        numerical_drift_method: str = 'KS',
+        categorical_drift_method: str = 'cramers_v',
+        ignore_na: bool = True,
+        aggregation_method: t.Optional[str] = 'l3_weighted',
+        min_samples: int = 10,
+        n_samples: int = 100_000,
+        random_state: int = 42,
+        **kwargs
+    ):
+        super().__init__(**kwargs)
+        self.properties = properties
+        self.ignore_properties = ignore_properties
+        self.n_top_properties = n_top_properties
+        self.margin_quantile_filter = margin_quantile_filter
+        self.max_num_categories_for_drift = max_num_categories_for_drift
+        self.min_category_size_ratio = min_category_size_ratio
+        self.max_num_categories_for_display = max_num_categories_for_display
+        self.show_categories_by = show_categories_by
+        self.numerical_drift_method = numerical_drift_method
+        self.categorical_drift_method = categorical_drift_method
+        self.ignore_na = ignore_na
+        self.aggregation_method = aggregation_method
+        self.min_samples = min_samples
+        self.n_samples = n_samples
+        self.random_state = random_state
+
+    def run_logic(self, context: Context) -> CheckResult:
+        """Run check."""
+        train = t.cast(TextData, context.train)
```

Comment:
why is the cast needed? can you do in the context property call?

Reply:
to say to the static type analyzer that the `context.train` is of type `TextData`

currently, the static type analyzer infers those properties to be of type `Never`, which means that code cannot and will not be reached, it is because `nlp.context.Context` type inherences those properties from `BaseContext` in which they are defined as `None`

```
class BaseContext(ABC):
    """Base class for contexts."""

    _train = None
    _test = None
    
    ...

    @property
    def train(self):
        """Return train if exists, otherwise raise error."""
        if self._train is None:
            raise DeepchecksNotSupportedError('Check is irrelevant for Datasets without train dataset')
        return self._train

```

![Never](https://user-images.githubusercontent.com/71635444/233051828-d77ae370-3863-40a5-92d8-efd816b710da.png)


> can you do in the context property call?

do you suggest overriding `train` and `test` properties in the nlp `Context` type?

Reply:
>can you do in the context property call?

or we can make `BaseContext` generic?

Reply:
I think the solution is too complicated so we can just leave it as is.
I thought we can do something simple like adding return type declaration to the property